src/notebookConcatDocument.ts

```diff
@@ -76,32 +87,102 @@ export class NotebookConcatDocument implements vscode.TextDocument, vscode.Dispo
+    private _realLines: NotebookConcatLine[] = [];
 
     public handleChange(e: protocol.TextDocumentEdit): protocol.DidChangeTextDocumentParams | undefined {
         this._version++;
         const changes: protocol.TextDocumentContentChangeEvent[] = [];
-        const cellIndex = this._cellRanges.findIndex((c) => c.uri.toString() === e.textDocument.uri);
-        const cell = cellIndex >= 0 ? this._cellRanges[cellIndex] : undefined;
-        if (cell) {
+        const index = this._spans.findIndex((c) => c.uri.toString() === e.textDocument.uri);
+        if (index >= 0) {
             e.edits.forEach((edit) => {
-                const normalized = edit.newText.replace(/\r/g, '');
-                const position = this.positionAt(cell.startOffset);
-                const from = new vscode.Position(position.line + edit.range.start.line, edit.range.start.character);
-                const to = new vscode.Position(position.line + edit.range.end.line, edit.range.end.character);
-                changes.push(...this.changeRange(normalized, from, to, cellIndex));
+                try {
+                    // Get old text for diffing
+                    const oldSpans = this._spans.filter((s) => s.uri.toString() === e.textDocument.uri);
+                    const oldText = oldSpans.map((s) => s.text).join('');
+
+                    // Apply the edit to the real spans
+                    const realText = this.getRealText(oldSpans[0].uri);
+                    const realCellLines = this._realLines.filter((r) => r.cellUri.toString() === e.textDocument.uri);
+                    const firstLineOffset = realCellLines[0].offset;
+                    const startOffset =
+                        realCellLines[edit.range.start.line].offset + edit.range.start.character - firstLineOffset;
+                    const endOffset =
+                        realCellLines[edit.range.end.line].offset + edit.range.end.character - firstLineOffset;
+                    const editedText = `${realText.slice(0, startOffset)}${edit.newText.replace(
+                        /\r/g,
+                        ''
+                    )}${realText.slice(endOffset)}`;
+
+                    // Create new spans from the edited text
+                    const newSpans = this.createSpans(
+                        oldSpans[0].uri,
+                        editedText,
+                        oldSpans[0].startOffset,
+                        oldSpans[0].realOffset
+                    );
+                    const newText = newSpans.map((s) => s.text).join('');
+
+                    // Diff the two pieces of text
+                    const diff = [...fastMyersDiff.diff(oldText, newText)];
+
+                    // Diff should have an array of numbers. These are
+                    // offsets in the old text translating to offsets in the new text
+                    if (diff.length > 0) {
+                        const oldTextStart = diff[0][0];
```

Comment:
Is there a way to maybe document what the fastMyersDiff returns? Even with the comment I'm not fully sure on what our diff indexes are here.

Reply:
Like having this in from the documentation on npm "it produces a sequence of quadruples [sx, ex, sy, ey], where [sx, ex) indicates a range to delete from xs and [sy, ey) indicates a range from ys to replace the deleted material with. Simple deletions are indicated when sy === ey and simple insertions when sx === ex"

Reply:
It does have docs? Do you mean put a link to the docs here? Otherwise the comment would be rather large.

Reply:
I put more comments and a link.

Reply:
Thanks that helps. It was reading a bit funny to me since it was looking at the individual diffs all as one big diff here (taking the start of the first diff to the end of the last diff) but that works for what this is doing.